[Durable Objects] Document new storage API improvements. (#2084)
cloudflare/cloudflare-docs: products/workers/src/content/learning/using-durable-objects.md

> <Aside>

Reviewer: ```suggestion
<Aside type="note">
```

> HTTP requests received by a Durable Object do not come directly from the Internet. They come from other Worker code – possibly other Durable Objects, or just plain Workers. We'll see how to send such a request in a bit. Durable Objects use HTTP for familiarity, but we plan to introduce other protocols in the future.

Author: Note: I changed all the `--`s to Unicode mdashes. GitHub's font here renders it like a regular hyphen, but I promise it's actually a wide dash.

Reviewer: I noticed it does this too in some previous, but thanks for the heads up :)

> HTTP requests received by a Durable Object do not come directly from the Internet. They come from other Worker code – possibly other Durable Objects, or just plain Workers. We'll see how to send such a request in a bit. Durable Objects use HTTP for familiarity, but we plan to introduce other protocols in the future.

Reviewer: ```suggestion
HTTP requests received by a Durable Object do not come directly from the Internet. They come from other Worker code – possibly other Durable Objects, or just plain Workers. You will learn how to send such a request later in this article. Durable Objects currently use HTTP for familiarity, with other protocols to be introduced in the future.
```

@kentonv can you please link me the part of the learning article specifically that references this sentence? It's best practice to hyperlink a section when referencing it in writing. This one: "We'll see how to send such a request in a bit."

}
```

The Durable Objects storage API employs several techniques to help you avoid subtle-yet-common storage bugs:

- Each individual storage operation is strictly ordered with respect to all others. Even if the operation completes asynchronously (requiring you to `await` a promise), the results will always be accuate as of the time the operation was invoked.

> - A Durable Object can process multiple concurrent requests. However, when a storage operation is in progress (such as, when you are `await`ing the result of a `get()`), delivery of concurrent events will be paused. This ensures that the state of the object cannot unexpectedly change while a read operation is in-flight, which would otherwise make it very hard to keep in-memory state properly synchronized with on-disk state. If desired, his behavior can be bypassed using the option `allowConcurrency: true`.

Reviewer 2: Can you link these mentions of options to the corresponding part of the runtime-apis doc?

> - If multiple writes operations are performed consecutively – without `await`ing anything in the meantime – then they will automatically be coalesced and applied atomically. This means that, even in the case of a machine failure, either all of the operations will have been stored to disk, or none of them will have been.

Reviewer: Was 'atomically' intentional or a spelling mistake?

> - If multiple writes operations are performed consecutively – without `await`ing anything in the meantime – then they will automatically be coalesced and applied atomically. This means that, even in the case of a machine failure, either all of the operations will have been stored to disk, or none of them will have been.

Reviewer: ```suggestion
- If multiple write operations are performed consecutively – without `await`ing anything in the meantime – then they will automatically be coalesced and applied atomically. This means that, even in the case of a machine failure, either all of the operations will have been stored to disk, or none of them will have been.
```

> - Write operations are queued to a write buffer, allowing calls like `put()` and `delete()` to complete "immediately" from the application's point of view. However, when the application initiates an outgoing network message (such as responding to a request, or invoking `fetch()`), the network request will be held until all previous writes are confirmed to be durable. This ensures that an application cannot accidentally confirm a write prematurely. If desired, his behavior can be bypassed using the option `allowUnconfirmed: true`.

Reviewer: ```suggestion
- Write operations are queued to a write buffer, allowing calls like `put()` and `delete()` to complete immediately from the application's point of view. However, when the application initiates an outgoing network message (such as responding to a request, or invoking `fetch()`), the network request will be held until all previous writes are confirmed to be durable. This ensures that an application cannot accidentally confirm a write prematurely. If desired, this behavior can be bypassed using the option `allowUnconfirmed: true`.
```

> - The storage API implements an in-memory caching layer to improve performance. Reads that hit cache will return instantly, without even context-switching to another thread. When reading or writing a value where caching is not worthwhile, you may use the option `noCache: true` to avoid it – but this option only affects performance, it won't change behavior.

Reviewer: ```suggestion
- The storage API implements an in-memory caching layer to improve performance. Reads that hit cache will return instantly, without even context-switching to another thread. When reading or writing a value where caching is not worthwhile, you may use the option `noCache: true` to avoid it – but this option only affects performance, it will not change behavior.
```

export class Counter {
    constructor(state, env) {
        this.state = state;
        // `blockConcurrencyWhile()` ensures no requests are delivered until
        // initialization completes.
        this.state.blockConcurrencyWhile(async () => {
            let stored = await this.state.storage.get("value");

>             // after initialization, future reads don't need to access storage!

Reviewer: ```suggestion
            // After initialization, future reads do not need to access storage.
```

            this.value = stored || 0;
        })
    }

    // Handle HTTP requests from clients.
    async fetch(request) {
        // this.value will retain its state until this object is evicted from memory
        ...
    }
}
```

<Aside header="Built-in Caching">

> The Durable Object's storage has a built-in in-memory cache of its own – if you `get()` a value that was read or written recently, the result will be instantly returned from cache. So, instead of writing initialization code like above, you could simply `get("value")` whenever you need it, and rely on the built-in cache to make this fast. However, in applications with more complex state, explicitly storing state in your object like above may be easier than making storage API calls on every access. We recommend writing your code in the way that's easiest for you.

Reviewer: ```suggestion
The Durable Object's storage has a built-in in-memory cache of its own – if you `get()` a value that was read or written recently, the result will be instantly returned from cache. So, instead of writing initialization code like above, you could simply `get("value")` whenever you need it, and rely on the built-in cache to make this fast. However, in applications with more complex state, explicitly storing state in your object like above may be easier than making storage API calls on every access. Depending on the configuration of your project, write your code in the way that is easiest for you.
```

</Aside>

### WebSockets in Durable Objects

> As part of Durable Objects, we've made it possible for Workers to act as WebSocket endpoints – including as a client or as a server. Previously, Workers could proxy WebSocket connections on to a back-end server, but could not speak the protocol directly.

Reviewer: ```suggestion
As part of Durable Objects, it is possible for Workers to act as WebSocket endpoints – including as a client or as a server. Previously, Workers could proxy WebSocket connections on to a back-end server, but could not speak the protocol directly.
```

Using Durable Objects will often add response latency, as the request must be forwarded to the point-of-presence where the object is located.

> While Durable Objects already perform well for many kinds of tasks, we have lots of performance tuning to do. Expect performance (latency, throughput, overhead, etc.) to improve over the beta period – and if you observe a performance problem, please tell us about it!

Reviewer: ```suggestion
While Durable Objects already perform well for many kinds of tasks, performance tuning is still expected. Performance (latency, throughput, overhead, etc.) will improve over the beta period – and if you observe a performance problem, [connect with the Workers community on Discord](https://discord.gg/cloudflaredev) or contact your account team to report your finding.
```
